File: frontend/src/lib/freshness.ts
/**
 * Where the data on screen actually came from, as reported by the backend.
 *
 * Every timestamp here is the server's, read off the response headers that
 * `freshness.py` attaches. The browser deliberately does not date anything
 * itself: a client clock records when the response *arrived*, which for
 * anything served from one of the app's cache layers is a different and much
 * more flattering number than when the data was pulled. It also cannot tell a
 * fresh answer from a week-old fallback, because both arrive as a normal 200.
 *
 * The store is keyed by request path and read by pathname, so a panel can ask
 * about `/stocks/AAPL` without knowing which query string the fetch used.
 */

import { useSyncExternalStore } from 'react'

export type DataSource = 'live' | 'cached' | 'archive' | 'stale'

export interface DataFreshness {
  source: DataSource
  /** When the backend pulled this from upstream. Null when unknowable. */
  fetchedAt: Date | null
  /** Newest data point in the payload (ISO date), for series like the chart. */
  dataThrough: string | null
  /** Server-computed age at response time, in seconds. */
  ageSeconds: number | null
  /** Client receipt time — used only to order entries, never displayed. */
  receivedAt: number
}

const store = new Map<string, DataFreshness>()
const listeners = new Set<() => void>()

// useSyncExternalStore compares snapshots by identity, so a mutable Map would
// never look changed. Bumping a counter on every write is the cheap way to
// give it something stable to diff.
let version = 0

function emit() {
  version += 1
  listeners.forEach(fn => fn())
}

/** Record what the backend said about a response. Called for every request. */
export function recordFreshness(path: string, res: Response): void {
  const source = res.headers.get('X-Data-Source') as DataSource | null
  // Absent on responses with nothing upstream behind them (/health, errors,
  // guest-mode short-circuits). Nothing to claim, so nothing is stored.
  if (!source) return

  const fetchedAtRaw = res.headers.get('X-Data-Fetched-At')
  const age = res.headers.get('X-Data-Age-Seconds')
  const parsed = fetchedAtRaw ? new Date(fetchedAtRaw) : null

  store.set(path, {
    source,
    fetchedAt: parsed && !Number.isNaN(parsed.getTime()) ? parsed : null,
    dataThrough: res.headers.get('X-Data-Through'),
    ageSeconds: age !== null ? Number(age) : null,
    receivedAt: Date.now(),
  })
  emit()
}

function pathnameOf(path: string): string {
  const q = path.indexOf('?')
  return q === -1 ? path : path.slice(0, q)
}

/**
 * Freshness for `key`, matching on the path and ignoring the query string.
 *
 * Two-step on purpose. An exact pathname match wins, so `/portfolio/` reports
 * the portfolio fetch rather than whichever `/portfolio/{ticker}/dividends`
 * call happened to land most recently. Only when nothing matches exactly does
 * it fall back to the newest entry under the prefix, which is what lets a
 * caller name a resource whose real path it doesn't assemble itself.
 */
export function readFreshness(key: string): DataFreshness | null {
  let exact: DataFreshness | null = null
  let prefixed: DataFreshness | null = null

  for (const [path, entry] of store) {
    const pathname = pathnameOf(path)
    if (pathname === key) {
      if (!exact || entry.receivedAt > exact.receivedAt) exact = entry
    } else if (pathname.startsWith(key)) {
      if (!prefixed || entry.receivedAt > prefixed.receivedAt) prefixed = entry
    }
  }
  return exact ?? prefixed
}

function subscribe(fn: () => void): () => void {
  listeners.add(fn)
  return () => { listeners.delete(fn) }
}

/** Live-updating freshness for the response matching `key`. */
export function useDataFreshness(key: string | null | undefined): DataFreshness | null {
  const snapshot = useSyncExternalStore(
    subscribe,
    () => (key ? `${version}:${key}` : ''),
    () => '',
  )
  // The subscription value is only a change token; the read happens here so
  // the returned object is always the current one.
  void snapshot
  return key ? readFreshness(key) : null
}

/** Test/sign-out hook: forget everything recorded so far. */
export function clearFreshness(): void {
  store.clear()
  emit()
}

// ── Presentation ────────────────────────────────────────────────────────────

/** "just now", "6m ago", "3h ago", "2d ago". */
export function formatAge(from: Date, now: number = Date.now()): string {
  const seconds = Math.max(0, Math.round((now - from.getTime()) / 1000))
  if (seconds < 45) return 'just now'
  const minutes = Math.round(seconds / 60)
  if (minutes < 60) return `${minutes}m ago`
  const hours = Math.round(minutes / 60)
  if (hours < 24) return `${hours}h ago`
  return `${Math.round(hours / 24)}d ago`
}

/** Exact server time, for the tooltip. Local timezone, seconds included. */
export function formatExact(at: Date): string {
  return at.toLocaleString(undefined, {
    year: 'numeric', month: 'short', day: 'numeric',
    hour: '2-digit', minute: '2-digit', second: '2-digit',
  })
}

export const SOURCE_LABEL: Record<DataSource, string> = {
  live: 'Live',
  cached: 'Cached',
  archive: 'Archived',
  stale: 'Stale',
}

export const SOURCE_BLURB: Record<DataSource, string> = {
  live: 'Fetched from the data provider while loading this view.',
  cached: 'Served from the server cache. It was pulled from the provider at the time shown.',
  archive: 'Read from the stored price history rather than fetched live.',
  stale: 'A previous copy — the live fetch failed (usually a provider rate limit), so the last good data is shown instead.',
}

File: backend/src/freshness.py
'''
Server-side provenance for every payload the app renders.

The problem this solves: almost nothing the frontend displays came from
upstream just now. A quote may be sixty seconds old, a fundamentals snapshot
six hours, a chart is read from an archive that a rate limit may have left a
week behind — and all four render identically. The user cannot tell current
data from a cached copy from a stale one being served because the live fetch
failed, which is exactly the distinction that decides whether they should act
on what they are looking at.

Timestamps are produced *here*, not in the browser. A frontend clock records
when the response arrived, which for anything cached at any of the app's five
layers is a different (and much more flattering) number than when the data was
actually pulled. It also can't see the difference between a fresh answer and a
week-old fallback, since both arrive as a normal 200.

Mechanics: a request-scoped ContextVar collects a stamp from each layer that
contributed, and middleware reports the *least fresh* of them on the response.
Least-fresh, not first or last, because an aggregate endpoint blends tiers —
/stocks/{ticker}/dashboard mixes a live quote with a six-hour snapshot and an
archived chart, and describing that as "live" because one part of it was is
the misreport this module exists to prevent.

Fail-soft throughout: a stamp that cannot be recorded costs a header, and no
caller should ever have to handle an error from annotating its own data.
'''

from __future__ import annotations

import logging
from contextvars import ContextVar
from dataclasses import dataclass
from datetime import date, datetime, timezone

logger = logging.getLogger(__name__)

# Sources, ordered by distance from live. The header reports the maximum.
LIVE = "live"          # fetched from upstream during this request
CACHED = "cached"      # served from a TTL cache that has not expired
ARCHIVE = "archive"    # read from a DB table (market_data, index_history, …)
STALE = "stale"        # a superseded copy, served because a live fetch failed

_RANK = {LIVE: 0, CACHED: 1, ARCHIVE: 2, STALE: 3}


@dataclass(frozen=True)
class Stamp:
    source: str
    fetched_at: datetime | None
    # Newest data point in the payload, for series that carry their own
    # calendar. A chart pulled two minutes ago whose last bar is from Friday
    # is fresh *and* three days behind; only reporting one of those numbers
    # hides the question the user is actually asking.
    data_through: date | None = None
    label: str | None = None


_stamps: ContextVar[list[Stamp] | None] = ContextVar("freshness_stamps", default=None)

# Distinguishes "didn't say when, so it's now" (a live fetch) from "genuinely
# don't know when this was pulled" — archive rows written before market_data
# tracked a fetch time. Collapsing the two would date every legacy bar to the
# moment it was read, which is the single most misleading thing this module
# could report.
_NOW = object()


def begin() -> None:
    '''Start collecting for a request. Called by the middleware.'''
    _stamps.set([])


def stamp(
    source: str,
    fetched_at: datetime | float | None = _NOW,
    data_through: date | str | None = None,
    label: str | None = None,
) -> None:
    '''Record that `source` contributed to the current response.

    `fetched_at` accepts a datetime or a POSIX timestamp (what the caches
    store). Omitting it means "now", which is only correct for a live fetch;
    passing None explicitly means the fetch time is genuinely unknown.
    '''
    collected = _stamps.get()
    if collected is None:
        return  # outside a request (background sweep, test, script)
    try:
        collected.append(Stamp(
            source=source,
            fetched_at=_now() if fetched_at is _NOW else (
                _as_datetime(fetched_at) if fetched_at is not None else None
            ),
            data_through=_as_date(data_through),
            label=label,
        ))
    except Exception as e:  # noqa: BLE001 — provenance must never break a response
        logger.debug("Could not record freshness stamp (%s): %r", source, e)


def snapshot() -> Stamp | None:
    '''The least-fresh contribution to this response, or None if unstamped.'''
    collected = _stamps.get()
    if not collected:
        return None
    worst = max(collected, key=lambda s: (_RANK.get(s.source, 0), -_epoch(s.fetched_at)))
    # The oldest fetch time and the oldest data point may come from different
    # stamps; report each one honestly rather than whichever happened to share
    # a record with the worst source.
    oldest_fetch = min((s.fetched_at for s in collected if s.fetched_at), default=None)
    through = min((s.data_through for s in collected if s.data_through), default=None)
    return Stamp(
        source=worst.source, fetched_at=oldest_fetch, data_through=through, label=worst.label,
    )


def _now() -> datetime:
    return datetime.now(timezone.utc)


def _epoch(value: datetime | None) -> float:
    return value.timestamp() if value else 0.0


def _as_datetime(value: datetime | float) -> datetime:
    if isinstance(value, datetime):
        return value if value.tzinfo else value.replace(tzinfo=timezone.utc)
    return datetime.fromtimestamp(float(value), tz=timezone.utc)


def _as_date(value: date | str | None) -> date | None:
    if value is None or isinstance(value, date):
        return value
    try:
        return date.fromisoformat(str(value)[:10])
    except ValueError:
        return None
